drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,224 @@
+	struct adc_keys_data *data = dev->data;
+	int ret;
+
+	if (!device_is_ready(cfg->channel.dev)) {
+		LOG_ERR("ADC controller device %s not ready", cfg->channel.dev->name);
+		return -ENODEV;
+	}
+
+	ret = adc_channel_setup_dt(&cfg->channel);
+	if (ret) {
+		LOG_ERR("ADC channel setup failed %d", ret);
+		return ret;
+	}
+
+	adc_sequence_init_dt(&cfg->channel, &data->seq);
+
+	data->self = dev;
+	k_work_init_delayable(&data->dwork, adc_keys_work_handler);
+
+#if defined(CONFIG_INPUT_LOG_LEVEL_DBG)
+	for (uint32_t i = 0; i < cfg->code_cnt; i++) {
+		LOG_DBG("* code %d: key_index=%d threshold=%d mV code=%d", i,
+			cfg->code_cfg[i].key_index, cfg->code_cfg[i].press_mv,
+			cfg->key_code[cfg->code_cfg[i].key_index]);
+	}
+#endif /* CONFIG_INPUT_LOG_LEVEL_DBG */
+
+	k_work_schedule(&data->dwork, K_MSEC(cfg->sample_period_ms));
+
+	return 0;
+}
+
+#define ADC_KEYS_CODE_CFG_CODE(node_id, prop, idx)                                                 \
+	{                                                                                          \
+		.key_index = DT_NODE_CHILD_IDX(node_id) /* include disabled nodes */,              \
+		.press_mv = DT_PROP_BY_IDX(node_id, prop, idx),                                    \
+	}
+
+#define ADC_KEYS_CODE_CFG(node_id)                                                                 \
+	DT_FOREACH_PROP_ELEM_SEP(node_id, press_thresholds_mv, ADC_KEYS_CODE_CFG_CODE, (,))
+
+#define ADC_KEYS_KEY_CODE(node_id) DT_PROP(node_id, zephyr_code)
+
+#define ADC_KEYS_INST(n)                                                                           \
+	static struct adc_keys_data adc_keys_data_##n;                                             \
+                                                                                                   \
+	static const struct adc_keys_code_config adc_keys_code_cfg_##n[] = {                       \
+		DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP(n, ADC_KEYS_CODE_CFG, (,))};                 \
+                                                                                                   \
+	static const uint16_t adc_keys_key_code_##n[] = {                                          \
+		DT_INST_FOREACH_CHILD_SEP(n, ADC_KEYS_KEY_CODE, (,))};                             \
+                                                                                                   \
+	static const struct adc_keys_config adc_keys_cfg_##n = {                                   \
+		.channel = ADC_DT_SPEC_INST_GET(n),                                                \
+		.sample_period_ms = DT_INST_PROP(n, sample_period_ms),                             \
+		.keyup_mv = DT_INST_PROP(n, keyup_threshold_mv),                                   \
+		.code_cfg = adc_keys_code_cfg_##n,                                                 \
+		.code_cnt = ARRAY_SIZE(adc_keys_code_cfg_##n),                                     \
+		.key_code = adc_keys_key_code_##n,                                                 \
```

Comment:
cool, this needs to initialize an array of `key_state` as well now, right?

Reply:
oops!